staging/src/k8s.io/apimachinery/pkg/util/validation/field/testing/match.go

```diff
@@ -138,66 +98,104 @@ func (m *Matcher) Render(e *field.Error) string {
+func (m ErrorMatcher) ByOrigin() ErrorMatcher {
 	m.matchOrigin = true
 	return m
 }
 
-// RequireOriginWhenInvalid configures the matcher to require the Origin field
-// to be set when the Type is Invalid and the matcher is matching by Origin.
-func (m *Matcher) RequireOriginWhenInvalid() *Matcher {
+// RequireOriginWhenInvalid returns a derived ErrorMatcher which also requires
+// the Origin field to be set when the Type is Invalid and the matcher is
+// matching by Origin.
+func (m ErrorMatcher) RequireOriginWhenInvalid() ErrorMatcher {
 	m.requireOriginWhenInvalid = true
 	return m
 }
 
-// ByDetailExact configures the matcher to match errors by the exact detail
-// string.
-func (m *Matcher) ByDetailExact() *Matcher {
+// ByDetailExact returns a derived ErrorMatcher which also matches errors by
+// the exact detail string.
+func (m ErrorMatcher) ByDetailExact() ErrorMatcher {
 	m.matchDetail = func(want, got string) bool {
 		return got == want
 	}
 	return m
 }
 
-// ByDetailSubstring configures the matcher to match errors by a substring of
-// the detail string.
-func (m *Matcher) ByDetailSubstring() *Matcher {
+// ByDetailSubstring returns a derived ErrorMatcher which also matches errors
+// by a substring of the detail string.
+func (m ErrorMatcher) ByDetailSubstring() ErrorMatcher {
 	m.matchDetail = func(want, got string) bool {
 		return strings.Contains(got, want)
 	}
 	return m
 }
 
-// ByDetailRegexp configures the matcher to match errors by a regular
-// expression of the detail string, where the "want" string is assumed to be a
-// valid regular expression.
-func (m *Matcher) ByDetailRegexp() *Matcher {
+// ByDetailRegexp returns a derived ErrorMatcher which also matches errors by a
+// regular expression of the detail string, where the "want" string is assumed
+// to be a valid regular expression.
+func (m ErrorMatcher) ByDetailRegexp() ErrorMatcher {
 	m.matchDetail = func(want, got string) bool {
 		return regexp.MustCompile(want).MatchString(got)
 	}
 	return m
 }
+
+// Test compares two ErrorLists by the criteria configured in this matcher, and
+// fails the test if they don't match. If a given "want" error matches multiple
+// "got" errors, they will all be consumed. This might be OK (e.g. if there are
+// multiple errors on the same field from the same origin) or it might be an
+// insufficiently specific matcher, so these will be logged.
+func (m ErrorMatcher) Test(t *testing.T, want, got field.ErrorList) {
+	t.Helper()
+
+	remaining := got
+	for _, w := range want {
+		tmp := make(field.ErrorList, 0, len(remaining))
+		n := 0
+		for _, g := range remaining {
+			if m.Matches(w, g) {
+				n++
+			} else {
+				tmp = append(tmp, g)
+			}
+		}
+		if n == 0 {
+			t.Errorf("expected an error matching:\n%s", m.Render(w))
+		} else if n > 1 {
+			// This is not necessarily and error, but it's worth logging in
+			// case it's not what the test author intended.
+			t.Logf("multiple errors matched:\n%s", m.Render(w))
```

Comment:
This discussion is still unresolved: https://github.com/kubernetes/kubernetes/pull/130388#discussion_r1977199117

This shouldn't block this PR too long because it is a useful enhancement. Feel free to lift the hold if you want to merge and iterate separately.

/lgtm
/hold

Reply:
https://github.com/kubernetes/kubernetes/pull/130388#discussion_r1978083357 resolves that thread.

Reply:
https://github.com/kubernetes/kubernetes/pull/130388#discussion_r1978092051 seems to resolve it?

Reply:
I was faster! :grin

/lgtm